urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+        }
+        diffs.add(new Diff(Operation.INSERT, param));
+        break;
+      case '-':
+        // Fall through.
+      case '=':
+        int n;
+        try {
+          n = Integer.parseInt(param);
+        } catch (NumberFormatException e) {
+          throw new IllegalArgumentException(
+              "Invalid number in diff_fromDelta: " + param, e);
+        }
+        if (n < 0) {
+          throw new IllegalArgumentException(
+              "Negative number in diff_fromDelta: " + param);
+        }
+        String text;
+        try {
+          text = text1.substring(pointer, pointer += n);
+        } catch (StringIndexOutOfBoundsException e) {
+          throw new IllegalArgumentException("Delta length (" + pointer
+              + ") larger than source text length (" + text1.length()
+              + ").", e);
+        }
+        if (token.charAt(0) == '=') {
+          diffs.add(new Diff(Operation.EQUAL, text));
+        } else {
+          diffs.add(new Diff(Operation.DELETE, text));
+        }
+        break;
+      default:
+        // Anything else is an error.
+        throw new IllegalArgumentException(
+            "Invalid diff operation in diff_fromDelta: " + token.charAt(0));
+      }
+    }
+    if (pointer != text1.length()) {
+      throw new IllegalArgumentException("Delta length (" + pointer
+          + ") smaller than source text length (" + text1.length() + ").");
+    }
+    return diffs;
+  }
+
+
+  //  MATCH FUNCTIONS
+
+
+  /**
+   * Locate the best instance of 'pattern' in 'text' near 'loc'.
+   * Returns -1 if no match found.
+   * @param text The text to search.
+   * @param pattern The pattern to search for.
+   * @param loc The location to search around.
+   * @return Best match index or -1.
+   */
+  public int match_main(String text, String pattern, int loc) {
+    // Check for null inputs.
+    if (text == null || pattern == null) {
+      throw new IllegalArgumentException("Null inputs. (match_main)");
+    }
+
+    loc = Math.max(0, Math.min(loc, text.length()));
+    if (text.equals(pattern)) {
+      // Shortcut (potentially not guaranteed by the algorithm)
+      return 0;
+    } else if (text.length() == 0) {
+      // Nothing to match.
+      return -1;
+    } else if (loc + pattern.length() <= text.length()
+        && text.substring(loc, loc + pattern.length()).equals(pattern)) {
+      // Perfect match at the perfect spot!  (Includes case of null pattern)
+      return loc;
+    } else {
+      // Do a fuzzy compare.
+      return match_bitap(text, pattern, loc);
+    }
+  }
+
+  /**
+   * Locate the best instance of 'pattern' in 'text' near 'loc' using the
+   * Bitap algorithm.  Returns -1 if no match found.
+   * @param text The text to search.
+   * @param pattern The pattern to search for.
+   * @param loc The location to search around.
+   * @return Best match index or -1.
+   */
+  protected int match_bitap(String text, String pattern, int loc) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.